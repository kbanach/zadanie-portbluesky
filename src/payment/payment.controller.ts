import {
  Controller,
  Get,
  HttpException,
  HttpStatus,
  Param,
  Query,
  Res,
} from "@nestjs/common";
import stringify from "csv-stringify/lib/sync";
import { Response } from "express";
import { PaymentFormat, PaymentDate, PaymentDatesQuery } from "./payment.dto";
import { PaymentService } from "./payment.service";

@Controller("v1/payment")
export class PaymentController {
  constructor(private readonly PaymentService: PaymentService) {}

  @Get(":paymentStartDate")
  getPaymentDates(
    @Param("paymentStartDate") startDate: string,
    @Query() query: PaymentDatesQuery,
    @Res() res: Response
  ) {
    try {
      const paymentDates = this.PaymentService.getPaymentDates(startDate);

      switch (query.format) {
        case PaymentFormat.CSV:
          return res.send(
            Buffer.from(this.PaymentService.paymentDatesToCSV(paymentDates))
          );
        case PaymentFormat.TEXT:
        case PaymentFormat.JSON:
        default:
          return res.json(paymentDates);
      }
    } catch (err) {
      throw new HttpException(err.message, HttpStatus.BAD_REQUEST);
    }
  }
}
