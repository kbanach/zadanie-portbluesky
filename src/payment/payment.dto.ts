export enum PaymentFormat {
  CSV = "csv",
  TEXT = "text",
  JSON = "JSON",
}

export interface PaymentDate {
  baseSalaryDate: string;
  bonusDate: string;
}

export interface PaymentDatesQuery {
  format: string;
}
