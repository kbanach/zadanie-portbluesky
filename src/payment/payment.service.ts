import { Injectable } from "@nestjs/common";
const stringify = require("csv-stringify/lib/sync");
import {
  addMonths,
  formatISO,
  isSaturday,
  isSunday,
  lastDayOfMonth,
  nextWednesday,
  parseISO,
  setDate,
  subDays,
} from "date-fns";
import { PaymentDate } from "./payment.dto";

@Injectable()
export class PaymentService {
  public readonly ISO_DATE_REGEX =
    /^\d{4}\-(0?[1-9]|1[012])\-(0?[1-9]|[12][0-9]|3[01])$/;

  private parseWithOffset(isoDate, monthOffset): Date {
    return addMonths(parseISO(isoDate), monthOffset);
  }

  private getNextBaseSalaryPaymentForDate(
    isoDate: string,
    monthOffset: number
  ): string {
    if (!this.ISO_DATE_REGEX.test(isoDate)) {
      throw new Error("Wrong date format. Please try YYYY-MM-DD");
    }

    let paymentDay = lastDayOfMonth(this.parseWithOffset(isoDate, monthOffset));

    if (isSaturday(paymentDay)) {
      paymentDay = subDays(paymentDay, 1);
    } else if (isSunday(paymentDay)) {
      paymentDay = subDays(paymentDay, 2);
    }

    return formatISO(paymentDay, { representation: "date" });
  }

  private getNextBonusPaymentForDate(
    isoDate: string,
    monthOffset: number
  ): string {
    if (!this.ISO_DATE_REGEX.test(isoDate)) {
      throw new Error("Wrong date format. Please try YYYY-MM-DD");
    }

    const parsedDate = this.parseWithOffset(isoDate, monthOffset);
    let paymentDay = setDate(parsedDate, 15);

    if (isSaturday(paymentDay) || isSunday(paymentDay)) {
      paymentDay = nextWednesday(paymentDay);
    }

    return formatISO(paymentDay, { representation: "date" });
  }

  public getPaymentDates(startDate: string, length = 12): PaymentDate[] {
    const paymentDates: PaymentDate[] = [];

    for (let i = 0; i < length; i++) {
      paymentDates.push({
        baseSalaryDate: this.getNextBaseSalaryPaymentForDate(startDate, i),
        bonusDate: this.getNextBonusPaymentForDate(startDate, i),
      });
    }

    return paymentDates;
  }

  private getCsvHeaders(): string[] {
    return ["Bonus payment date", "Base salary payment date"];
  }

  private paymentDateToCsvRow(date: PaymentDate): string[] {
    return [date.bonusDate, date.baseSalaryDate];
  }

  public paymentDatesToCSV(paymentDates: PaymentDate[]): string {
    return stringify([
      this.getCsvHeaders(),
      ...paymentDates.map((d) => this.paymentDateToCsvRow(d)),
    ]);
  }
}
