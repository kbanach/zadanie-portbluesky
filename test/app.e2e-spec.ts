import { Test, TestingModule } from "@nestjs/testing";
import { INestApplication } from "@nestjs/common";
import * as request from "supertest";
import { PaymentModule } from "./../src/payment/payment.module";

describe("AppController (e2e)", () => {
  let app: INestApplication;

  beforeEach(async () => {
    const moduleFixture: TestingModule = await Test.createTestingModule({
      imports: [PaymentModule],
    }).compile();

    app = moduleFixture.createNestApplication();
    await app.init();
  });

  describe("/v1/payment/2021-01-01 (GET)", () => {
    it("returns 200", () => {
      return request(app.getHttpServer())
        .get("/v1/payment/2021-01-01")
        .expect(200);
    });

    it("returns proper dates for 2021-01-01", () => {
      return request(app.getHttpServer())
        .get("/v1/payment/2021-01-01")
        .expect((res) => {
          expect(res.body).toMatchSnapshot();
        });
    });

    it("returns CSV file for parameter format=csv for 2021-01-01", () => {
      return request(app.getHttpServer())
        .get("/v1/payment/2021-01-01?format=csv")
        .expect((res) => {
          expect(res.headers["content-type"]).toMatch(
            "application/octet-stream"
          );
          expect(res.body).toMatchSnapshot();
        });
    });
  });
});
