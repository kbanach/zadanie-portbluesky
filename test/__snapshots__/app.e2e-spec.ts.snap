// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppController (e2e) /v1/payment/2021-01-01 (GET) returns CSV file for parameter format=csv for 2021-01-01 1`] = `
Object {
  "data": Array [
    66,
    111,
    110,
    117,
    115,
    32,
    112,
    97,
    121,
    109,
    101,
    110,
    116,
    32,
    100,
    97,
    116,
    101,
    44,
    66,
    97,
    115,
    101,
    32,
    115,
    97,
    108,
    97,
    114,
    121,
    32,
    112,
    97,
    121,
    109,
    101,
    110,
    116,
    32,
    100,
    97,
    116,
    101,
    10,
    50,
    48,
    50,
    49,
    45,
    48,
    49,
    45,
    49,
    53,
    44,
    50,
    48,
    50,
    49,
    45,
    48,
    49,
    45,
    50,
    57,
    10,
    50,
    48,
    50,
    49,
    45,
    48,
    50,
    45,
    49,
    53,
    44,
    50,
    48,
    50,
    49,
    45,
    48,
    50,
    45,
    50,
    54,
    10,
    50,
    48,
    50,
    49,
    45,
    48,
    51,
    45,
    49,
    53,
    44,
    50,
    48,
    50,
    49,
    45,
    48,
    51,
    45,
    51,
    49,
    10,
    50,
    48,
    50,
    49,
    45,
    48,
    52,
    45,
    49,
    53,
    44,
    50,
    48,
    50,
    49,
    45,
    48,
    52,
    45,
    51,
    48,
    10,
    50,
    48,
    50,
    49,
    45,
    48,
    53,
    45,
    49,
    57,
    44,
    50,
    48,
    50,
    49,
    45,
    48,
    53,
    45,
    51,
    49,
    10,
    50,
    48,
    50,
    49,
    45,
    48,
    54,
    45,
    49,
    53,
    44,
    50,
    48,
    50,
    49,
    45,
    48,
    54,
    45,
    51,
    48,
    10,
    50,
    48,
    50,
    49,
    45,
    48,
    55,
    45,
    49,
    53,
    44,
    50,
    48,
    50,
    49,
    45,
    48,
    55,
    45,
    51,
    48,
    10,
    50,
    48,
    50,
    49,
    45,
    48,
    56,
    45,
    49,
    56,
    44,
    50,
    48,
    50,
    49,
    45,
    48,
    56,
    45,
    51,
    49,
    10,
    50,
    48,
    50,
    49,
    45,
    48,
    57,
    45,
    49,
    53,
    44,
    50,
    48,
    50,
    49,
    45,
    48,
    57,
    45,
    51,
    48,
    10,
    50,
    48,
    50,
    49,
    45,
    49,
    48,
    45,
    49,
    53,
    44,
    50,
    48,
    50,
    49,
    45,
    49,
    48,
    45,
    50,
    57,
    10,
    50,
    48,
    50,
    49,
    45,
    49,
    49,
    45,
    49,
    53,
    44,
    50,
    48,
    50,
    49,
    45,
    49,
    49,
    45,
    51,
    48,
    10,
    50,
    48,
    50,
    49,
    45,
    49,
    50,
    45,
    49,
    53,
    44,
    50,
    48,
    50,
    49,
    45,
    49,
    50,
    45,
    51,
    49,
    10,
  ],
  "type": "Buffer",
}
`;

exports[`AppController (e2e) /v1/payment/2021-01-01 (GET) returns proper dates for 2021-01-01 1`] = `
Array [
  Object {
    "baseSalaryDate": "2021-01-29",
    "bonusDate": "2021-01-15",
  },
  Object {
    "baseSalaryDate": "2021-02-26",
    "bonusDate": "2021-02-15",
  },
  Object {
    "baseSalaryDate": "2021-03-31",
    "bonusDate": "2021-03-15",
  },
  Object {
    "baseSalaryDate": "2021-04-30",
    "bonusDate": "2021-04-15",
  },
  Object {
    "baseSalaryDate": "2021-05-31",
    "bonusDate": "2021-05-19",
  },
  Object {
    "baseSalaryDate": "2021-06-30",
    "bonusDate": "2021-06-15",
  },
  Object {
    "baseSalaryDate": "2021-07-30",
    "bonusDate": "2021-07-15",
  },
  Object {
    "baseSalaryDate": "2021-08-31",
    "bonusDate": "2021-08-18",
  },
  Object {
    "baseSalaryDate": "2021-09-30",
    "bonusDate": "2021-09-15",
  },
  Object {
    "baseSalaryDate": "2021-10-29",
    "bonusDate": "2021-10-15",
  },
  Object {
    "baseSalaryDate": "2021-11-30",
    "bonusDate": "2021-11-15",
  },
  Object {
    "baseSalaryDate": "2021-12-31",
    "bonusDate": "2021-12-15",
  },
]
`;
